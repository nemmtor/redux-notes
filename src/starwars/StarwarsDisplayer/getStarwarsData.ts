export interface ResponseJson {
  name: string;
  height: string;
  mass: string;
  hair_color: string;
  skin_color: string;
  eye_color: string;
  birth_year: string;
  gender: string;
  homeworld: string;
  films: string[];
  species: unknown[];
  vehicles: string[];
  starships: string[];
  created: Date;
  edited: Date;
  url: string;
}

export const getStarwarsData = async (): Promise<ResponseJson> => {
  const response = await fetch('https://swapi.dev/api/people/1');
  const json = await response.json();
  return json;
};
