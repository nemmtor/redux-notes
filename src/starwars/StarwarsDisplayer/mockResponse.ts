export const mockResponse = {
  name: 'luke',
  surname: 'skywalker',
};
