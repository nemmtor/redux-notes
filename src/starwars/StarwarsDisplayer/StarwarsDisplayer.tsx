import React, { useState } from 'react';
import { getStarwarsData, ResponseJson } from './getStarwarsData';

export const StarwarsDisplayer: React.FC = () => {
  const [info, setInfo] = useState<ResponseJson | null>();
  const handleClick = async () => {
    const fetchedInfo = await getStarwarsData();
    setInfo(fetchedInfo);
  };

  return (
    <main data-testid="starwars-test">
      <button data-testid="starwars-button" type="button" onClick={handleClick}>
        Fetch starwars info
      </button>
      {info && (
        <pre data-testid="starwars-output">{JSON.stringify(info, null, 4)}</pre>
      )}
    </main>
  );
};
