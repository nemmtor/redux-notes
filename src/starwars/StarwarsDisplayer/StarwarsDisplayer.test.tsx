import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';

import { StarwarsDisplayer } from './StarwarsDisplayer';
import { getStarwarsData } from './getStarwarsData';
import { mockResponse } from './mockResponse';

jest.mock('./getStarwarsData');

const getStarwarsDataMock = getStarwarsData as jest.Mock;

describe('NewNoteInput', () => {
  it('has valid text content', () => {
    const { getByTestId } = render(<StarwarsDisplayer />);
    expect(getByTestId('starwars-test')).toHaveTextContent(
      'Fetch starwars info',
    );
  });

  it('calls func on click', () => {
    const { getByTestId } = render(<StarwarsDisplayer />);
    const button = getByTestId('starwars-button');
    fireEvent.click(button);
    expect(getStarwarsData).toBeCalled();
  });

  it('outputs data from fetch after clicking', async () => {
    getStarwarsDataMock.mockResolvedValueOnce(mockResponse);
    const { getByTestId, findByTestId } = render(<StarwarsDisplayer />);
    const button = getByTestId('starwars-button');
    fireEvent.click(button);
    const output = await findByTestId('starwars-output');
    expect(output).toBeInTheDocument();
    expect(output.textContent).toEqual(JSON.stringify(mockResponse, null, 4));
  });

  it('matches snapshot', () => {
    const tree = renderer.create(<StarwarsDisplayer />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
