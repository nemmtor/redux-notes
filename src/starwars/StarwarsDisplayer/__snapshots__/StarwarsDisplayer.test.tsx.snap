// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewNoteInput matches snapshot 1`] = `
<main
  data-testid="starwars-test"
>
  <button
    data-testid="starwars-button"
    onClick={[Function]}
    type="button"
  >
    Fetch starwars info
  </button>
</main>
`;
